importer/trickle/trickledag.go

```diff
@@ -56,3 +58,210 @@ func fillTrickleRec(db *h.DagBuilderHelper, node *h.UnixfsNode, depth int) error
 	}
 	return nil
 }
+
+// TrickleAppend appends the data in `db` to the dag, using the Trickledag format
+func TrickleAppend(base *dag.Node, db *h.DagBuilderHelper) (*dag.Node, error) {
+	// Convert to unixfs node for working with easily
+	ufsn, err := h.NewUnixfsNodeFromDag(base)
+	if err != nil {
+		return nil, err
+	}
+
+	// Get depth of this 'tree'
+	n, j := trickleDepthInfo(ufsn, db.Maxlinks())
+	if n == 0 {
+		// If direct blocks not filled...
+		err := db.FillNodeLayer(ufsn)
+		if err != nil {
+			return nil, err
+		}
+
+		if db.Done() {
+			return ufsn.GetDagNode()
+		}
+
+		// If continuing, our depth has increased by one
+		n++
+	}
+
+	err = appendFillLastChild(ufsn, n-1, j, db)
+	if err != nil {
+		return nil, err
+	}
+
+	// after appendFillLastChild, our depth is now increased by one
+	if !db.Done() {
+		n++
+	}
+
+	// Now, continue filling out tree like normal
+	for i := n; !db.Done(); i++ {
+		for j := 0; j < layerRepeat && !db.Done(); j++ {
+			next := h.NewUnixfsNode()
+			err := fillTrickleRec(db, next, i)
+			if err != nil {
+				return nil, err
+			}
+
+			err = ufsn.AddChild(next, db)
+			if err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	return ufsn.GetDagNode()
+}
+
+func appendFillLastChild(ufsn *h.UnixfsNode, depth int, layerFill int, db *h.DagBuilderHelper) error {
+	if ufsn.NumChildren() > db.Maxlinks() {
+		// Recursive step, grab last child
+		last := ufsn.NumChildren() - 1
+		lastChild, err := ufsn.GetChild(last, db.GetDagServ())
+		if err != nil {
+			return err
+		}
+
+		// Fill out last child (may not be full tree)
+		nchild, err := trickleAppendRec(lastChild, db, depth-1)
+		if err != nil {
+			return err
+		}
+
+		// Update changed child in parent node
+		ufsn.RemoveChild(last)
+		err = ufsn.AddChild(nchild, db)
+		if err != nil {
+			return err
+		}
+
+		// Partially filled depth layer
+		if layerFill != 0 {
+			for ; layerFill < layerRepeat && !db.Done(); layerFill++ {
+				next := h.NewUnixfsNode()
+				err := fillTrickleRec(db, next, depth)
+				if err != nil {
+					return err
+				}
+
+				err = ufsn.AddChild(next, db)
+				if err != nil {
+					return err
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+func trickleAppendRec(ufsn *h.UnixfsNode, db *h.DagBuilderHelper, depth int) (*h.UnixfsNode, error) {
```

Comment:
my guess is there's a recast of `TrickleAppend`, `appendFillLastChild`, and `trickleAppendRec` that shares a lot more code and is very very simple (like the original construction).